Add OpenVINO backend

wine99 opened the pull request:
### Overview

This PR introduces an [OpenVINO backend](https://docs.openvino.ai/2025/index.html) for `llama.cpp`, enabling hardware-accelerated inference on **Intel® CPUs, GPUs, and NPUs**. The backend leverages OpenVINO to deliver optimized inference with the existing llama.cpp GGUF model ecosystem. Enables performance improvements via OpenVINO’s graph compilation and kernel fusion.

* llama.cpp with OpenVINO backend: [Build Instructions](https://github.com/ravi9/llama.cpp/blob/dev_backend_openvino/docs/build.md#openvino)

### Key Features:

* **New backend implementation**
  * Added OpenVINO backend in `ggml/src/ggml-openvino`.
  * Implemented translations for core GGML operations

* **Supported precisions**
  * FP16/BF16 GGUF models supported.
  * Q4_0, Q4_1, Q4_K_M, Q6_K models partially supported. (See notes below)

* **Supported devices**
  * Intel CPUs
  * Intel integrated and discrete GPUs
  * Intel NPUs (requires **UD32+ driver**).

**For NPU: currently prompt processing is slow, a smaller context size is recommended for better performance, e.g., `-c 512`.**

**For llama-bench: `-fa 1` is required.**

### Tested Models

The following models are validated for functionality.

Accuracy and performance are WIP.

* [`Llama-3.2-1B-Instruct-GGUF`](https://huggingface.co/MaziyarPanahi/Llama-3.2-1B-Instruct-GGUF)
* [`Llama-3.1-8B-Instruct`](https://huggingface.co/meta-llama/Llama-3.1-8B-Instruct) 
* [`microsoft/Phi-3-mini-4k-instruct-gguf`](https://huggingface.co/microsoft/Phi-3-mini-4k-instruct-gguf)
* [`Qwen/Qwen2.5-1.5B-Instruct-GGUF`](https://huggingface.co/Qwen/Qwen2.5-1.5B-Instruct-GGUF)
* [`Qwen/Qwen3-8B`](https://huggingface.co/Qwen/Qwen3-8B)
* [`openbmb/MiniCPM-1B-sft-bf16`](https://huggingface.co/openbmb/MiniCPM-1B-sft-bf16)
* [`tencent/Hunyuan-7B-Instruct`](https://huggingface.co/tencent/Hunyuan-7B-Instruct)
* [`mistralai/Mistral-7B-Instruct-v0.3`](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3)

### Work in Progress
* Performance and memory optimizations 
* Broader quantization coverage.
* Support for additional model architectures. 
* Extensive accuracy testing.

### Notes on quantization support

#### CPU
* **Q4_0, Q4_1, Q4_K_M and Q6_K models are supported.**
* Q6_K tensors (6bit gs16 sym) are converted to int8 gs16 sym.
* Q5_K tensors (5bit gs32 asym) are converted to int8 gs32 asym.

#### GPU
* **Q4_0, Q4_1, Q4_K_M and Q6_K models are supported.**
* Q6_K tensors (6bit gs16 sym) are requantized to int8 gs32 sym.
* Q5_K tensors (5bit gs32 asym) are converted to int8 gs32 asym.

#### NPU
* **Main quantization scheme for the supported models in this PR is Q4_0.**
* Q4_0 and Q4_1 tensors are requantized to int4 gs128 sym.
* Q6_K tensors are dequantized to fp16.

Other notes:
* Both Q4_0 and Q4_1 models use Q6_K for the token_embedding tensor and the weight tensor in the last matmul (in most models it is the same tensor as token_emb).
* Q4_0 models will produce some Q4_1 tensors if imatrix is provided as part of the quantization of the model using llama-quantize utility.
* Q4_K_M models additionally have Q6_K tensors and Q5_K tensors (only in Phi3 in the validated model list of this PR).

NOTE: Optimum-intel converts the fp16/bf16 token embedding tensor and the weight tensor in the last matmul to int8 asym channel-wise ([config code](https://github.com/huggingface/optimum-intel/blob/b60e4d4866509a1aeea2b7a3f26f2a70bc464354/optimum/commands/export/openvino.py#L183-L191)).

wine99 committed:
Update build doc

wine99 committed:
Add cgraph tensor output name to OV op name

wine99 committed:
Add initial NPU support

ravi9 committed:
Update openvino build instructions

wine99 committed:
draft NPU support version 2: prefill + kvcache

wine99 committed:
NPU support version 2: prefill + kvcache

wine99 committed:
Change due to ggml cgraph changes, not correct yet

wine99 committed:
Change due to ggml cgraph changes, llama-3.2 CPU work

wine99 committed:
Add AMD64 to CMakeLists

wine99 committed:
Change due to ggml cgraph changes, all device work

wine99 committed:
Refactor: clean, fix warning

wine99 committed:
Update clang-format

wine99 committed:
Statful transformation for CPU GPU

wine99 committed:
Add SwiGLU

wine99 committed:
Fuse to SDPA

wine99 committed:
Replace Concat with Broadcast in MulMat for GQA

wine99 committed:
Pull out indices creation for kv cache update

wine99 committed:
Refactor: remove past_token_len from extra_inputs

wine99 committed:
Fix Phi3 SwiGLU and SoftMax

wine99 committed:
Pull out sin cos from rope

wine99 committed:
Reduce memory: free ov weights node after graph conversion

wine99 committed:
Fix CPY due to cgraph change

ravi9 committed:
Added OpenVINO CI/CD. Updated docs

wine99 committed:
Fix Phi3 ROPE; Add test-backend-ops

wine99 committed:
Fix llama-cli

wine99 committed:
Fix NPU

wine99 committed:
Fix llama-bench; Clang-format

wine99 committed:
Fix llama-perplexity

cavusmustafa committed:
temp. changes for mark decomp

wine99 committed:
matmul in fp32

cavusmustafa committed:
mulmat input conversion fix

cavusmustafa committed:
mulmat type conversion update

cavusmustafa committed:
add mark decomp pass

wine99 committed:
Revert changes in fuse_to_sdpa

ravi9 committed:
Update build.md

wine99 committed:
Fix test-backend-ops

wine99 committed:
Skip test-thread-safety; Run ctest only in ci/run.sh

wine99 committed:
Use CiD for NPU

wine99 committed:
Optimize tensor conversion, improve TTFT

wine99 committed:
Update supports_buft and supports_op for quantized models

wine99 committed:
Add quant weight conversion functions from genai gguf reader

wine99 committed:
Quant models run with accuracy issue

wine99 committed:
Fix accuracy: disable cpu_repack

wine99 committed:
Fix CI; Disable test-backend-ops

wine99 committed:
Fix Q4_1

wine99 committed:
Fix test-backend-ops: Treat quantized tensors as weights

wine99 committed:
Support op SET_ROWS

wine99 committed:
Fix NPU

wine99 committed:
Remove CPY

wine99 committed:
Fix test-backend-ops

wine99 committed:
Minor updates for raising PR

wine99 committed:
Add NPU Q4_0 support

ravi9 commented:
Hi @SearchSavior ,
>Q: Will this feature in llama.cpp offer wider GGUF coverage via something like parameter mapping?

Instead of converting GGUF models to PyTorch format with parameter mapping, this implementation uses OpenVINO's GGML frontend to directly translate GGML computation graphs to OpenVINO operations at runtime. The translation happens through a comprehensive operation mapping system that covers the core GGML operations. Since it works at the GGML operation level, it should support any model architecture that llama.cpp supports (assuming we map/translate all the GGML operators to OpenVINO.) 
 
>Q: What parts of the OpenVINO feature set are intended to be brought into llama.cpp?

The immediate focus is on runtime acceleration: kernel fusion, optimized graph execution,memory optimizations and hardware scheduling on CPU, GPU, and NPU.
 
>Q: Is this PR trying to bring in only performance from openvino runtime to support llama.cpp usecase?

The scope of this PR is primarily performance enablement using OpenVINO runtime to accelerate llama.cpp inference while preserving compatibility with the GGUF ecosystem. It’s not introducing a new model conversion flow, so everything remains driven by GGUF models in llama.cpp.
 
>Q: Will pipeline parallel / heterogeneous execution be supported here?

We are currently reviewing this. llama.cpp already has infrastructure for pipeline parallelism, and the OpenVINO backend exposes async operations and events, so it should be possible. Further evaluation is needed to confirm integration details.

wine99 committed:
NPU perf: eliminate zp

Bionic-Squash commented:
I can't wait for openVINO support to get upstreamed

wine99 committed:
Perf: RMS fused to OV internal RMS op

wine99 committed:
Dequantize q4_1 q4_k q6_k for NPU

wine99 committed:
Add custom quant type: q8_1_c, q4_0_128

wine99 committed:
Set m_is_static=false as default in decoder

wine99 committed:
Simpilfy translation of get_rows

wine99 committed:
Fix after rebasing

- Layout of cache k and cache v are unified: [seq, n_head, head_size]
- Add CPY and FLASH_ATTN_EXT, flash attn is not used yet
- Skip test-backend-ops due to flash attn test crash
- Add mutex around graph conversion to avoid test-thread-safety fali in the future
- Update NPU config
- Update GPU config to disable SDPA opt to make phi-3 run

wine99 committed:
Change openvino device_type to GPU; Enable flash_attn

wine99 committed:
Fix after rebasing

wine99 committed:
Improve debug util; Eliminate nop ReshapeReshape

wine99 committed:
STYLE: make get_types_to_requant a function

wine99 committed:
Support BF16 model

wine99 committed:
Fix NPU compile

wine99 committed:
WA for npu 1st token acc issue

wine99 committed:
Apply EliminateZP only for npu

wine99 committed:
Add GeGLU

wine99 committed:
Fix Hunyuan

wine99 committed:
Support iSWA

wine99 committed:
Fix NPU accuracy

wine99 committed:
Fix ROPE accuracy when freq_scale != 1

wine99 committed:
Minor: not add attention_size_swa for non-swa model

wine99 committed:
Minor refactor

wine99 committed:
Add Q5_K to support phi-3-q4_k_m

wine99 committed:
Requantize Q6_K (gs16) to gs32 on GPU

wine99 committed:
Fix after rebasing

wine99 committed:
Always apply Eliminate_ZP to fix GPU compile issue on some platforms

cavusmustafa committed:
kvcachefusion support

cavusmustafa committed:
env variable GGML_OPENVINO_DISABLE_SDPA_OPTIMIZATION added

cavusmustafa committed:
Fix for Phi3

wine99 committed:
Fix llama-cli (need to run with --no-warmup)

wine99 committed:
Fix add_sliced_mask; Revert mulmat, softmax; Remove input attention_size, iSWA model not working

wine99 committed:
fix after rebasing

wine99 committed:
Fix llama-3-8b and phi3-mini q4_0 NPU

ravi9 committed:
Update to OV-2025.3 and CMakeLists.txt

wine99 committed:
Add OV CI cache

CISC commented on .github/actions/linux-setup-openvino/action.yml:
```diff
@@ -0,0 +1,30 @@
+      with:
+        url: https://storage.openvinotoolkit.org/repositories/openvino/packages/${{ inputs.version_major }}/linux/openvino_toolkit_ubuntu24_${{ inputs.version_full }}_x86_64.tgz
+        path: ${{ inputs.path }}
+        type: "z"
```
```suggestion
        type: z
```

CISC commented on .github/workflows/build-cache.yml:
```diff
@@ -63,6 +63,39 @@ jobs:
+      - name: Dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y libtbb12
```
Is this needed for the installer?

CISC commented on .github/workflows/release.yml:
```diff
@@ -241,6 +241,63 @@ jobs:
+          fetch-depth: 0
+
+      - name: ccache
+        uses: hendrikmuhs/ccache-action@v1.2.16
```
```suggestion
        uses: ggml-org/ccache-action@v1.2.16
```

CISC commented on .github/workflows/release.yml:
```diff
@@ -241,6 +241,63 @@ jobs:
+      - name: Build
+        id: cmake_build
+        run: |
+          source /opt/intel/openvino/setupvars.sh
```
Use the cached SDK.

ravi9 committed:
Apply CISC review and update CI to OV2025.3

CISC commented on .github/actions/linux-setup-openvino/action.yml:
```diff
@@ -0,0 +1,30 @@
+      shell: bash
+      run: |
+        cd ${{ inputs.path }}
+        echo "Y" | sudo -E ./install_dependencies/install_openvino_dependencies.sh
```
This won't work if this installs anything outside `inputs.path`.

If it does you either have to cache that too somehow or run this in the `Build` step.

ravi9 replied:
Added as a step before `Build`

ravi9 commented:
@slaren Thanks for testing.
- We are working on fixing `llama-bench`.
- For `llama-cli` and `llama-server`, please run with `--no-warmup` for now. The input shapes for the warmup need to be fixed. We are working on a solution for it. `llama-cli -m models/Llama-3.2-1B-Instruct.Q4_K_M.gguf --no-warmup`
- `llama-simple` should work fine.

ravi9 committed:
Update CI to run OV dep install before build

ravi9 committed:
Update OV dockerfile to use OV2025.3 and update build docs

wine99 committed:
Style: use switch in supports_ops

wine99 commented on ggml/src/ggml-openvino/ggml-openvino.cpp:
```diff
@@ -0,0 +1,579 @@
+}
+
+static bool is_op_unsupported_case(const ggml_tensor* op) {
+    if (op->op == GGML_OP_SOFT_MAX) {
```
Updated with your suggestions.

wine99 commented on ggml/src/ggml-openvino/ggml-openvino.cpp:
```diff
@@ -0,0 +1,579 @@
+        float scale = 1.0f;
+        float max_bias = 0.0f;
+        float logit_softcap = 0.0f;
+        const auto* op_params = op->op_params;
```
Is middle pointer alignment (and reference alignment) the recommended format for backends? We added a clang-format rule enforcing left pointer alignment at some point, so most pointers are currently left aligned. If middle alignment is preferred, we can update all pointers to use it. Otherwise, we can change the few middle-aligned ones to left alignment for consistency.

danbev replied:
@wine99 Yes I believe this is the [preferred](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md#coding-guidelines) alignment even though I don't see it explicitly mentioned in the coding guidelines, most of the code base follows this 👍

wine99 replied:
Middle aligned all pointers and references. Also omitted optional struct keywords according to the guidelines.

Vertical alignment is not applied. Should that also be enforced?

wine99 committed:
Clean placeholders in ggml-openvino.cpp

wine99 committed:
Style: middle ptr and ref align, omit optional struct keyword

wine99 committed:
NPU Unify PD (#14)

* Stateless. Fix llama-cli llama-server

* Simplify broadcast op in attention

* Replace get_output_tensor+memcpy with set_output_tensor

* NPU unify PD. Unify dynamic and static dims

wine99 committed:
NPU unify PD (handled internally)

wine99 committed:
change graph to 4d, support multi sequences

wine99 committed:
Fix llama-bench

I-N-T-E-L committed:
Update ggml-decoder.cpp

Hitting error while compiling on windows:

error C3861: 'unsetenv': identifier not found

Reason: unsetenv() is a POSIX function; it doesn’t exist on Windows. Visual Studio (MSVC) won’t recognize it.

Proposed fix: Use _putenv_s() (Windows equivalent)
This is supported by MSVC and achieves the same effect: it removes the environment variable from the process environment.

This keeps cross-platform compatibility.

I-N-T-E-L committed:
Update ggml-decoder.cpp

I-N-T-E-L committed:
Update ggml-decoder.cpp

I-N-T-E-L committed:
Update ggml-decoder.cpp

I-N-T-E-L committed:
Update ggml-decoder.cpp

I-N-T-E-L committed:
Update ggml-decoder.cpp

wine99 committed:
Fix NPU

zhaixuejun1993 committed:
Remove the second decoder for node. Moving the function into the model decoder

zhaixuejun1993 committed:
Fix error for naive

wine99 committed:
fallback naive run with accuracy issue

wine99 committed:
NPU prefill chunking

wine99 committed:
NPU fix llama-bench

wine99 committed:
NPU support llma-perplexity -b 512 --no-warmup

zhaixuejun1993 committed:
remove unused API GgmlOvDecoder::get_output_stride(const std::string & name)

wine99 committed:
Refactor: split ov_graph_compute for dynamic and static

zhaixuejun1993 committed:
remove unused API GgmlOvDecoder::get_output_names()

zhaixuejun1993 committed:
remove unused API get_output_shape(const std::string & name)

zhaixuejun1993 committed:
Modified API GgmlOvDecoder::get_output_type(const std::string & name)

zhaixuejun1993 committed:
Removed API GgmlOvDecoder::get_output_op_params(const std::string & name)

zhaixuejun1993 committed:
Removed API get_output_ggml_tensor(const std::string & name)

zhaixuejun1993 committed:
Removed API m_outputs

zhaixuejun1993 committed:
Removed m_output_names

wine99 committed:
minor update due to ov 2025.4

zhaixuejun1993 committed:
Removed API GgmlOvDecoder::get_input_names()

zhaixuejun1993 committed:
Removed API GgmlOvDecoder::get_input_stride(const std::string& name)

zhaixuejun1993 committed:
Removed API get_input_type

zhaixuejun1993 committed:
Removed API get_input_type

zhaixuejun1993 committed:
Removed API GgmlOvDecoder::get_input_shape(const std::string & name)

zhaixuejun1993 committed:
Removed API GgmlOvDecoder::get_input_op_params(const std::string & name)

zhaixuejun1993 committed:
Fix error for decoder cache

wine99 committed:
Reuse cached decoder

wine99 committed:
GPU remove Q6_K requantization

wine99 committed:
NPU fix wrong model output shape

wine99 committed:
NPU fix q4 perf regression

zhaixuejun1993 committed:
Remove unused variable nodes

wine99 committed:
Fix decoder can_reuse for llama-bench

wine99 committed:
backend buffer: allocate on host

wine99 committed:
Use shared_buffer for GPU NPU; Refactor

wine99 committed:
Add ov_backend_host_buffer; Use cached remote context

wine99 committed:
Put kvcache on GPU

wine99 committed:
Use ggml_aligned_malloc

wine99 committed:
only use remote tensor for kvcache

wine99 committed:
only use remote tensor for kvcache for GPU

wine99 committed:
FIX: use remote tensor from singleton

I-N-T-E-L committed:
Update build.md for Windows

wine99 committed:
Update build.md to include OpenCL

wine99 committed:
NPU always requant to q4_0_128

wine99 committed:
Optimize symmetric quant weight extraction: use single zp

wine99 committed:
Use Q8_0_C in token embd, lm_head, and for 5 and 6 bits quant

wine99 committed:
Update build.md

yangsu2022 commented on ggml/src/ggml-openvino/utils.cpp:
```diff
@@ -0,0 +1,773 @@
+    return config;
+}
+
+std::map<ggml_type, ExtraQuantType> get_types_to_requant(const std::string & device) {
```
Could you confirm if GPU/CPU re-quantization has been removed? If so, please update the intro description for this PR.

wine99 replied:
will update PR description later

wine99 committed:
Support -ctk f32

cavusmustafa committed:
Initial stateful graph support

cavusmustafa committed:
Update ggml/src/ggml-openvino/ggml-decoder.cpp

Co-authored-by: Yamini Nimmagadda <yamini.nimmagadda@intel.com>

cavusmustafa committed:
code cleanup

cavusmustafa committed:
npu perf fix

cavusmustafa committed:
requant to f16 for Q6 embed on NPU

ynimmaga committed:
Create OPENVINO.md in llama.cpp backend docs

ynimmaga committed:
Update OPENVINO.md

ynimmaga committed:
Update OPENVINO.md

ynimmaga committed:
Update OPENVINO.md

ynimmaga committed:
Update build.md

cavusmustafa committed:
Update ggml/src/ggml-openvino/ggml-decoder.cpp

cavusmustafa committed:
Update ggml/src/ggml-openvino/ggml-openvino-extra.cpp

ynimmaga committed:
Update OPENVINO.md

ynimmaga committed:
Update OPENVINO.md

ynimmaga committed:
Update OPENVINO.md

cavusmustafa commented:
Hello,
Following the recent merge with llama.cpp master, we have encountered some breaking changes in the OpenVINO backend. We are currently adapting the code to these updates. We will update once adjustments are complete and it is ready for re-review.

cavusmustafa committed:
kq_mask naming fix

cavusmustafa committed:
Syntax correction for workflows build file

wine99 committed:
Change ov backend buffer is_host to false

YenFuLin commented:
Amazing work! Cannot wait this PR.
However, now I tried to run Llama-3.1-8B-Instruct-Q4_K_M on GPU and got nonsense response. But, it works on CPU.
Could someone help me to check this abnormal result ;(

## Env
OS: Win11
CPU: 11th Gen Intel(R) Core(TM) i5-11400
GPU: Intel(R) Arc(TM) B580 Graphics
OpenVINO: v2025.3.0
Model card: [unsloth/Llama-3.1-8B-Instruct-GGUF](https://huggingface.co/unsloth/Llama-3.1-8B-Instruct-GGUF)

## Command
```
set GGML_OPENVINO_DEVICE=GPU
llama-simple -m "R:\Meta-Llama-3.1-8B-Instruct-Q4_K_M.gguf" -n 50 "The story of AI is "
```

## Log
[GPU.txt](https://github.com/user-attachments/files/24764473/GPU.txt)
[CPU.txt](https://github.com/user-attachments/files/24764474/CPU.txt)

cavusmustafa committed:
Simplification for stateful and update output shape processing

wine99 committed:
Fix llama-bench -p -n where p<=256

wine99 committed:
Fix --direct-io 0

wine99 committed:
Remove hardcode names

wine99 committed:
Fix stateful shapes

wine99 committed:
Don't put kvcache on GPU in stateful mode

wine99 committed:
Remove hardcode names

wine99 committed:
Avoid re-compilation in llama-bench

wine99 committed:
Extract zp directly instead of bias

wine99 committed:
Refactor weight tensor processing

wine99 committed:
Merge branch 'master' into dev_backend_openvino

wine99 committed:
create_weight_node accept non-ov backend buffer

wine99 committed:
remove changes in llama-graph.cpp

cavusmustafa committed:
stateful masking fix (#38)

Fix for stateful accuracy issues and cl_out_of_resources error in stateful GPU with larger context sizes.

wine99 committed:
Fix test-backend-ops crash glu, get_rows, scale, rms_norm, add

cavusmustafa committed:
hardcoded name handling for rope_freqs.weight

wine99 committed:
Suppress logging and add error handling to allow test-backend-ops to complete

wine99 committed:
Fix MUL_MAT with broadcast; Add unsupported MUL_MAT FLASH_ATTN cases

wine99 committed:
Use bias instead of zp in test-backend-ops

cavusmustafa committed:
Merge pull request #43 from cavusmustafa/additional_fixes_after_rebase

hardcoded name handling for rope_freqs.weight

cavusmustafa commented on ggml/src/ggml-openvino/ggml-decoder.cpp:
```diff
@@ -0,0 +1,885 @@
+                compute_params.output_len = 1;
+            }
+        }
+    }
```
Thank you for the feedback on this, we updated the code to remove hardcoded tensor names.

ggerganov commented on ggml/src/ggml-openvino/utils.cpp:
```diff
@@ -0,0 +1,802 @@
+    static std::unordered_map<graph_key, std::shared_ptr<GgmlOvDecoder>, graph_key_hash> decoder_cache;
+    static std::unordered_map<graph_key, std::shared_ptr<ov::InferRequest>, graph_key_hash> infer_request_cache;
+    static std::unordered_map<graph_key, std::vector<std::string>, graph_key_hash> ov_input_names_cache;
+    static std::unordered_map<graph_key, std::vector<std::string>, graph_key_hash> ov_output_names_cache;
```
Having `static` data like this is a bad practice. Most likely these have to be moved to the backend context.

cavusmustafa replied:
Thank you for the feedback. We updated the code and removed static data.

ggerganov commented on ggml/src/ggml-openvino/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+find_package(OpenVINO REQUIRED)
+find_package(OpenCL REQUIRED)
+
+include("${OpenVINO_DIR}/../3rdparty/tbb/lib/cmake/TBB/TBBConfig.cmake")
```
Is this correct? It looks like there is implicit assumption that there will be a `3rdparty` directory next to OpenVINO dir. How do we know this is going to be satisfied?

ravi9 replied:
Yes, with OpenVINO archive installation, there will always be 3rdparty directory which is packaged with OpenVINO archive.

ravi9 committed:
Update OV in CI, Add OV CI Tests in GH Actions

CISC commented on .devops/openvino.Dockerfile:
```diff
@@ -0,0 +1,134 @@
+ARG OPENVINO_VERSION_MAJOR=2025.3
+ARG OPENVINO_VERSION_FULL=2025.3.0.19807.44526285f24
```
```suggestion
ARG OPENVINO_VERSION_MAJOR=2026.1
ARG OPENVINO_VERSION_FULL=2026.0.1.20960.91859257636
```

CISC commented on .github/workflows/build.yml:
```diff
@@ -743,6 +743,68 @@ jobs:
+          id: depends
+          run: |
+            sudo apt-get update
+            sudo apt-get install -y build-essential libcurl4-openssl-dev libtbb12 cmake ninja-build python3-pip
```
```suggestion
            sudo apt-get install -y build-essential libssl-dev libtbb12 cmake ninja-build python3-pip
```

cavusmustafa committed:
Temp fix for multithreading bug

ravi9 committed:
Update OV CI, fix review suggestions.

ravi9 committed:
Merge pull request #45 from cavusmustafa/tmp_fix_multithread

Temp fix for multithreading bug

ravi9 committed:
fix editorconfig-checker, update docs

ravi9 committed:
Fix tabs to spaces for editorconfig-checker

ravi9 committed:
fix editorconfig-checker

ravi9 committed:
Update docs

cavusmustafa committed:
updated model link to be GGUF model links

wine99 committed:
Remove GGML_CPU_REPACK=OFF

wine99 committed:
Merge branch 'master' into dev_backend_openvino

wine99 committed:
Skip permuted ADD and MUL

cavusmustafa committed:
Removed static variables from utils.cpp

cavusmustafa committed:
Removed initializing non-existing variable

wine99 committed:
Remove unused structs

cavusmustafa committed:
Merge pull request #1 from wine99/remove_static_variables

Remove unused structs

wine99 committed:
Removed static variables from utils.cpp

wine99 committed:
Fix test-backend-ops for OV GPU

zhaixuejun1993 committed:
unify api calling

zhaixuejun1993 committed:
Update utils.cpp

zhaixuejun1993 committed:
When the dim is dynamic, throw an error, need to is stastic forst

wine99 committed:
Fix test-backend-ops for OV GPU LNL

zhaixuejun1993 committed:
Add interface compute_model_outputs(), which get the model output through computing the node use count & status in the cgraph to avoid the flag using

zhaixuejun1993 committed:
No need to return

wine99 committed:
Merge branch 'master' into dev_backend_openvino

wine99 committed:
Fix test-thread-safety

Commit:
use the shape from infer request of output tensor create to avoid issue

Commit:
fix dynamic output shape  issue

wine99 committed:
Merge pull request #49 from zhaixuejun1993/xuejun/unify-api-get_ov_output_tensor

Commit:
fix issue for the unused node in tests

lcretan commented:
Hello,

I would like to ask the SYCL build maintainers that the build for SYCL includes OpenVINO as default, because this build already removed the supports of the other architectures like CUDA and this build is dedicated only for Intel XPUs.

In the Whisper.cpp project, OpenVINO is already officially supported and it is found that OpenVINO can accelerates CPU processing, not only GPU and NPU. 

It is also said that the combination of OneBLAS and OpenVINO is more that the sum of them. 

I hope that positively.
Thank you in advance.

wine99 committed:
Rewrite the logistic about model outputs computer, add new API compute_model_outputs()

wine99 committed:
Remove unused lock

wine99 committed:
Merge branch 'dev_backend_openvino' into fix-thread-safety

wine99 committed:
Fix test-thread-safety

wine99 committed:
Add comment

wine99 committed:
Fix test-backend-ops for OV GPU LNL

Temp fix test-backend-ops for OV GPU LNL

ravi9 committed:
Merge pull request #46 from cavusmustafa/fix-readme-model-links

updated model links to be GGUF model links

ravi9 committed:
Update openvino docs

ravi9 committed:
update to OV release version 2026.0

ravi9 committed:
add ci ov-gpu self hosted runner

ravi9 committed:
fix editorconfig

wine99 committed:
Fix perplexity

zhaixuejun1993 committed:
Put the iteration logistic in func

zhaixuejun1993 committed:
Rewrite the model inputs finding mechanism  (#54)

* Rewrite the model inputs finding logistic

* Put stateful shape handle in get input shape

cavusmustafa commented on ggml/src/ggml-openvino/ggml-decoder.cpp:
```diff
@@ -0,0 +1,960 @@
+std::map<std::string, std::shared_ptr<ov::Node>> GgmlOvDecoder::create_weight_nodes(ggml_cgraph * cgraph, bool naive) {
+    static std::mutex weights_mutex;
+    std::lock_guard<std::mutex> lock(weights_mutex);
+
```
We added this due to a failing thread-safety test. We are currently working on an alternative solution and will remove it once the code is updated accordingly.

ravi9 committed:
Added ggml-ci-intel-openvino-gpu and doc update

cavusmustafa committed:
.hpp files converted to .h

cavusmustafa committed:
Merge pull request #57 from cavusmustafa/hpp_to_h

.hpp files converted to .h

ravi9 committed:
fix ggml-ci-x64-intel-openvino-gpu

cavusmustafa committed:
Fix for stateful execution bug in llama-bench

cavusmustafa committed:
Minor updates after stateful llama-bench fix

cavusmustafa committed:
Update ggml/src/ggml-openvino/utils.cpp

Co-authored-by: Yamini Nimmagadda <yamini.nimmagadda@intel.com>

cavusmustafa committed:
Remove multiple get_shape calls

cavusmustafa committed:
Bring back mutex into compute

ggerganov commented on .github/workflows/build.yml:
```diff
@@ -743,6 +743,82 @@ jobs:
+              openvino_device: "GPU"
+
+      runs-on: ${{ fromJSON(matrix.runner) }}
+
```
Looks like the `llama-embedding` test is failing:

https://github.com/ggml-org/llama.cpp/actions/runs/22787048288/job/66120793212?pr=15307#step:6:3479

zhaixuejun1993 committed:
Fix VIEW op, which slice the input node

zhaixuejun1993 committed:
Added token_len_per_seq existence check before slicing masks and moved node retrieval inside guarded block to prevent missing-key access

cavusmustafa committed:
Merge pull request #60 from zhaixuejun1993/xuejun/hot-fix-llama-embedding

Fix VIEW op, which slice the input node

cavusmustafa committed:
Temp. fix for test requant errors

cavusmustafa committed:
Merge pull request #62 from zhaixuejun1993/xuejun/fix_issue_key_miss

Fix missing issue key handling

cavusmustafa committed:
Merge pull request #58 from cavusmustafa/fix_stateful_state_sync

Fix for stateful execution bug in llama-bench

ravi9 committed:
Update to OV ggml-ci to low-perf

ggerganov committed:
ci : temporary disable "test-llama-archs"

ggerganov committed:
ci : cache v4 -> v5, checkout v4 -> v6, fix runner tag

ggerganov committed:
docs : update url

ggerganov reviewed (approved):
@ravi9 I've disabled the `test-llama-archs` test as it was failing. Recommend to try to get it fixed as a priority because this will increase confidence that the OpenVINO backend is general enough to support all LLMs.

Waiting for CI to pass and merging.

ravi9 committed:
Fix OV link in docker and Update docs

ravi9 commented:
Hi @WizardlyBump17 ,
Apologies for the lack of clarity.
GPU drivers need to be installed inside the container to be able to detect GPU.
We plan to include GPU and NPU drivers inside the docker image in the next PR. You can use this 

[Update] Added [openvino.Dockerfile](https://github.com/ggml-org/llama.cpp/blob/master/.devops/openvino.Dockerfile)

You could also verify if the GPU is detected by running `clinfo` inside the container.

savvadesogle commented:
Hi @ravi9 
I may be wrong, but isn't installing just  intel-opencl-icd enough to detect the GPU in a container?

savvadesogle commented:
<img width="1396" height="768" alt="pr-15307" src="https://github.com/user-attachments/assets/f6a76433-0049-478e-b37d-c163d1299936" />

Thanks to Zijun Yu, Ravi Panchumarthy, Su Yang, Mustafa Cavus, Arshath, Xuejun Zhai, Yamini Nimmagadda, and Wang Yang, you've done such a great job!

And thanks to reviewers Sigbjørn Skjæret, Georgi Gerganov, and Daniel Bevenius for their strict supervision!

And please don't be offended if I missed anyone, you're all amazing!!!
